va/caa.go

```diff
@@ -116,7 +117,56 @@ func parseResults(results []caaResult) (*CAASet, error) {
 	return nil, nil
 }
 
-func (va *ValidationAuthorityImpl) parallelCAALookup(ctx context.Context, name string, lookuper func(context.Context, string) ([]*dns.CAA, error)) []caaResult {
+func withParents(fqdn string) []string {
+	var result []string
+	labels := strings.Split(fqdn, ".")
+	for i := 0; i < len(labels); i++ {
+		result = append(result, strings.Join(labels[i:], "."))
+	}
+	return result
+}
+
+// Implement pre-erratum 5065 style tree-climbing CAA. Note: a strict
+// interpretation of pre-5065 indicates a linear lookup path - if there is any
+// CNAME at all, that precludes further tree-climbing on the original FQDN. This
+// is clearly wrong. We implement a hybrid approach that is strictly more
+// conservative: We always do full tree-climbing on the original FQDN (by virtue
+// of parallelCAALookup. When the LegacyCAA flag is enabled, we also
+// do linear tree climbing on single-level aliases.
+func (va *ValidationAuthorityImpl) treeClimbingLookupCAA(ctx context.Context, fqdn string) ([]*dns.CAA, error) {
+	// We will do a maximum of 8 tree-climbing queries to avoid CNAME/CAA hybrid
+	// loops
+	maxAttempts := 8
+	return va.treeClimbingLookupCAAWithCount(ctx, fqdn, &maxAttempts)
+}
+
+func (va *ValidationAuthorityImpl) treeClimbingLookupCAAWithCount(ctx context.Context, fqdn string, maxAttempts *int) ([]*dns.CAA, error) {
+	if *maxAttempts < 1 {
+		return nil, nil
+	}
+	*maxAttempts--
+	caas, cnames, err := va.dnsClient.LookupCAA(ctx, fqdn)
+	if err != nil {
+		return nil, err
+	} else if len(caas) > 0 {
+		return caas, nil
+	} else if len(cnames) > 0 {
+		for _, cname := range cnames {
```

Comment:
If this only used the first CNAME (the target of the record with fqdn as the header name) I think we would avoid the repeated tree descension problem no?

i.e. we would go from
```
lookup caa(cname.example.com) = cname to cname.other.com, cname to final.com
       target = cname.other.com, lookup caa(cname.example.com) = cname to final.com
              target = (final.com), lookup caa(final.com) = empty
                     target = parent(final.com) = com, lookup caa(com) = empty
              target = parent(cname.other.com) = other.com, lookup caa(other.com) = empty
                     target = parent(other.com) = com, lookup caa(com) = empty
       target = final.com, lookup caa(final.com) = empty
              target = parent(final.com) = com, lookup caa(com) = empty
       target = parent(cname.example.com) = example.com, lookup caa(example.com) = empty
              target = parent(example.com) = com, lookup(com) = empty
```

to

```
lookup caa(cname.example.com) = cname to cname.other.com, cname to final.com
       target = cname.other.com, lookup caa(cname.example.com) = cname to final.com
              target = (final.com), lookup caa(final.com) = empty
                     target = parent(final.com) = com, lookup caa(com) = empty
              target = parent(cname.other.com) = other.com, lookup caa(other.com) = empty
                     target = parent(other.com) = com, lookup caa(com) = empty
       target = parent(cname.example.com) = example.com, lookup caa(example.com) = empty
              target = parent(example.com) = com, lookup(com) = empty
```

Reply:
basically this: https://github.com/letsencrypt/boulder/compare/legacy-caa...legacy-caa-opt?expand=1

Reply:
I wound up solving this a different way in the latest commit.